src/models.py

```diff
@@ -45,38 +98,78 @@ def __init__(self, num_hidden, embedding_dim, vocab_size, device, num_layers=2,
+        self.fc = nn.Linear(self.num_hidden * 2, vocab_size)
     
-    def forward(self, x, h0, seq_lengths):
-        # x has shape (batch, seq_len), h0 has shape (num_layers * num_directions, batch, num_hidden)
+    def forward(self, caption, h0, c0, seq_lengths, intermediate_layer):
+        # -------------- Inputs --------------------
+            # caption = caption; shape = (batch, seq_len)
+            # mlp = image feature vector averaged and passed through MLP
+            # seq_lengths = ''
+            # intermediate_layer = output of 'encoder.layer4[2].bn2' 
         
-        c0 = self.init_cell_state(h0.shape[1])
-        
-        emb = self.emb(x)  # (batch, seq_len) -> (batch, seq_len, embedding_dim)
+        # Initialising the hidden state and cell state of the LSTM with 
+        # the two different MLPs
+
+        c0, h0 = self.init_cell_state_and_hidden_state(caption.shape[0], c0, h0)
         
+        emb = self.emb(caption)  # (batch, seq_len) -> (batch, seq_len, embedding_dim)
+     
         # Pack_padded_sequence so that padded items in the sequence won't be shown to the LSTM
         # See https://towardsdatascience.com/taming-lstms-variable-sized-mini-batches-and-why-pytorch-is-good-for-your-health-61d35642972e
         # We use seq_lengths - 1 because we do not want to count the last <END> token in each caption.
         emb = torch.nn.utils.rnn.pack_padded_sequence(emb, seq_lengths - 1, batch_first=True)
+
+        # [2, 38, 512]
+
+        # input of shape (seq_len, batch, input_size):
         
         out, _ = self.lstm(emb, (h0, c0))
 
         # Undo the packing operation
         out, _ = torch.nn.utils.rnn.pad_packed_sequence(out, batch_first=True)  # (batch, seq_len, num_directions * num_hidden)
         
-        logits = self.fc(out)
+        # ---------------- Attention -----------------------
+
+        # torch.Size([2, 512, 16]) -> torch.Size([2, 16, 512])
+        intermediate_layer = torch.transpose(intermediate_layer, 1, -1)
+        
+        attn_in = torch.zeros(out.size(0), out.size(1), out.size(2) * 2).to(self.device)
+        
+        for i in range(out.size(0)):
+            seq_len = seq_lengths[i].item() - 1
+
+            stack_intermediate_layer = torch.zeros(out.size(1), out.size(2)).to(self.device)
+
+            # As the two tensors (out, intermediate_layer) have different shapes,
+            # I copy from intermediate_layer until I have reached shape [2, 38, 512] ('out' shape)
+            for j in range(seq_len):
+                idx = j % (intermediate_layer.size(1) - 1)
+                stack_intermediate_layer[j] = intermediate_layer[i][idx].clone().detach()
+
+            attn_in[i, :seq_len] = torch.cat((stack_intermediate_layer, out[i, :seq_len]), dim = 1)
+
+        attn_out = self.attn(attn_in)
+        attn_weights = F.softmax(attn_out, dim=1)
+        
+        cs = torch.sum(attn_weights.unsqueeze(1) * out, dim=1, keepdim=False)
```

Comment:
Once again, pay attention to the dimension. Are you not summing over the LSTM timesteps?

Reply:
Should it be

```suggestion
        cs = torch.sum(attn_weights.unsqueeze(1) * out, dim=2, keepdim=False)
```
because we are summing over the hidden layers? I tried that, but got a bug for some reason.

Reply:
Yes, I indeed meant it should be dim=2